### PR TITLE
Adds configuration API and retry policy for temp data

### PR DESCRIPTION
**Pull request by ig-sinicyn**
This PR attempts to add extensibility/configuration API.

Scenario: as a customer of the library I wish to be able to override default behavior for some features. Candidates are: logging, retry policy for temp data and so on.

Previously we had configuration API for code assertions only (see `CodeExceptionsHelper`). This PR adds TempData retry policy configuration. In future, there may be more extensibility points.

As a starting point, I've added single `CodeJam.Internal.Configuration` class that contains all configurable features. Personally, I do not like current design as it puts a lot of uncoupled code into single class. I cannot find a good alternative, so feedback is welcome!

As addition, I've added TimeoutHelper.ExponentialBackoffTimeout method that is used by default temp data retry policy implementation.

### CODE REVIEW

File: CodeJam.Main.Tests/Threading/TimeoutHelperTests.cs [hunk 1]
```diff
@@ -74,5 +76,22 @@ public void TestAdjustTimeoutLimitInfiniteIfDefault()
...
+		[TestCase(10, 60)]
+		[TestCase(100, 60)]
+		public void TestBackoff(int retry, int resultSeconds)
+		{
+			var delay = TimeSpan.FromSeconds(1);
+			var max = TimeSpan.FromMinutes(1);
+			var timeoutSeconds = TimeoutHelper.ExponentialBackoffTimeout(retry, delay, max).TotalSeconds;
+			var range = Range.Create(resultSeconds * 0.8, resultSeconds * 1.2);
+
+			Assert.IsTrue(range.Contains(timeoutSeconds), $"{range} does not contain {timeoutSeconds}");
```
Comment (NN---) [inline CodeJam.Main.Tests/Threading/TimeoutHelperTests.cs:94]: Better to use CollectionAssert.Contains or Assert.That(, Does.Contain(..))

Comment (ig-sinicyn) [reply CodeJam.Main.Tests/Threading/TimeoutHelperTests.cs:94]: Range is a special type, CodeJam.Range and it does not implement ICollection/IEnumerable.

So there's no way to use it together with NUnit assertion methods.

Comment (ig-sinicyn) [reply CodeJam.Main.Tests/Threading/TimeoutHelperTests.cs]: ... so I 've replaced it with Assert ... Is.InRange() :)

File: CodeJam.Main.Tests/IO/TempDataTests.cs [hunk 2]
```diff
@@ -98,6 +112,46 @@ public void TestDirectoryNestedContent()
...
+						Assert.AreEqual(disposeCallCount, 1);
+					}
+
+					Assert.DoesNotThrow(() => dir.Dispose());
+					Assert.AreEqual(disposeCallCount, 2);
+
+				}
+				Assert.IsFalse(Directory.Exists(dirPath), "Directory should NOT exist");
+				Assert.IsFalse(File.Exists(nestedFile), "File should NOT exist");
+				Assert.AreEqual(disposeCallCount, 2);
```
Comment (NN---) [inline CodeJam.Main.Tests/IO/TempDataTests.cs:147]: 2 should come first.
Order is expected , actual.

File: CodeJam.Main.Tests/IO/TempDataTests.cs [hunk 3]
```diff
@@ -98,6 +112,46 @@ public void TestDirectoryNestedContent()
...
+					nestedFile = Path.Combine(dirPath, "test.tmp");
+
+					using (File.Create(nestedFile))
+					{
+						Assert.Throws<IOException>(() => dir.Dispose());
+						Assert.AreEqual(disposeCallCount, 1);
+					}
+
+					Assert.DoesNotThrow(() => dir.Dispose());
+					Assert.AreEqual(disposeCallCount, 2);
```
Comment (NN---) [inline CodeJam.Main.Tests/IO/TempDataTests.cs:142]: 2 comes first

File: CodeJam.Main.Tests/IO/TempDataTests.cs [hunk 4]
```diff
@@ -98,6 +112,46 @@ public void TestDirectoryNestedContent()
...
+
+				}
+				Assert.IsFalse(Directory.Exists(dirPath), "Directory should NOT exist");
```
Comment (NN---) [inline CodeJam.Main.Tests/IO/TempDataTests.cs:145]: Use the better constraint ;)

https://github.com/nunit/docs/wiki/FileOrDirectoryExistsConstraint

File: CodeJam.Main.Tests/IO/TempDataTests.cs (hunk 3 above)
Comment (ig-sinicyn) [reply CodeJam.Main.Tests/IO/TempDataTests.cs:142]: nope, Dispose() call increments disposeCallCount variable

File: CodeJam.Main.Tests/IO/TempDataTests.cs (hunk 4 above)
Comment (ig-sinicyn) [reply CodeJam.Main.Tests/IO/TempDataTests.cs:145]: Well, there's almost half handred Exists() calls in the file and I'm not going to rewrite them all:)

Second, I'd prefer to keep direct framework calls. Proposed one may become controversial very quickly. As example, 
`Assert.That(somevar, Does.NotExists)`.

File: CodeJam.Main.Tests/IO/TempDataTests.cs (hunk 3 above)
Comment (NN---) [reply CodeJam.Main.Tests/IO/TempDataTests.cs:142]: I mean

Assert.AreEqual(2, disposeCallCount);

File: CodeJam.Main.Tests/IO/TempDataTests.cs (hunk 2 above)
Comment (ig-sinicyn) [reply CodeJam.Main.Tests/IO/TempDataTests.cs]: Fixed)

File: CodeJam.Main.Tests/IO/TempDataTests.cs (hunk 4 above)
Comment (NN---) [reply CodeJam.Main.Tests/IO/TempDataTests.cs:146]: What is the problem ?

Assert.That(foolder, Does.Not.Exist)


And you can combine

Assert.That(foolder, Is.Not.Null.And.Does.Not.Exist)

I’ll check how we can replace all exists with this constraint

File: CodeJam.Main.Tests/IO/TempDataTests.cs (hunk 3 above)
Comment (ig-sinicyn) [reply CodeJam.Main.Tests/IO/TempDataTests.cs:142]: nope too. We're using standard order of arguments (`Compare(actual, expected)`) in all tests.

So, there're two options:
1. Rewrite all tests assertions.
2. Do not bother with awkward nUnit design decisions and keep current code as is. 

I strongly prefer second one as it has a very good value to effort ratio :)

**Comment by ig-sinicyn** [comment]
@NN--- , thanks for the review 👍

**Comment by ig-sinicyn** [comment]
+1 for @andrewvk :)

**Pull request merged**